Using pundit to ensure all controller actions have authorisation policy coverage

christophwong opened the pull request:
Move `after_action :verify_authorized` from documents_controller to
application_controller. This allows pundit to throw a warning when we
expose a controller action without proper authorisation rules in place.
Pass tests by adding policy for all controller actions

The pundit policy related to manuals are completely open right now, this has no effect on the existing authorisations and scoping that is inside of controller actions for manuals. The pundit policy offers some boilerplate for when we want to implement all the rules there.

Commit:
[RED TESTS]Add user authorisation to attachments_controller

Move `after_action :verify_authorized` from documents_controller to
application_controller. This allows pundit to throw a warning when we
expose a controller action without proper authorisation rules in place.

We extract ApplicationPolicy as the parent class to reduce duplicate
helper methods. However, we allow duplicate policy-rule methods so it
is easier to read all the rules for each controller in one place.

This commit is failing some tests but those shall be fix with the next
few commits when policies are implemented for other controllers.

Commit:
Pass tests by adding policy for all controller actions

The policy related to manuals are completely open right now, but there
are some permissions/scoping in place such as the code in
`ManualsController#index`. We decided to first do a generic rule to
quiet down pundit's errors, and when we play the card for manuals
permissions, we can implement authorisation rules for manuals.

There was also a change in routes, because we had been quietly passing
through the `DocumentController#index` actions which redirects us back
to `ManualsController#index` when we go to root.

rgarner commented on app/policies/manual_policy.rb:
```diff
@@ -0,0 +1,11 @@
+  alias_method :create?, :index?
+  alias_method :edit?, :index?
+  alias_method :update?, :index?
+  alias_method :show?, :index?
```
If the intent is to allow all these actions by default, would something like

``` ruby
[:index?, :new?, :create?, :edit?, :update?, :show?].each do |allowed_action|
  define_method allowed_action do
    true
  end
end
```

communicate the same intent without indicating via `alias_method` that the methods are semantically equivalent (and that `:index?` has some kind of primary status)?

benilovj replied:
@rgarner I take your point about `:index?` having some kind of primary status, but I personally find the `alias_method` approach slightly easier to parse visually than the meta-programming approach (and the semantic equivalence doesn't offend me as much). This is all very subjective though.

I think this may be moot because we'll likely revise `ManualPolicy` to reflect the SPv1 rules tomorrow.

rgarner commented on app/policies/application_policy.rb:
```diff
@@ -0,0 +1,23 @@
+  def gds_editor
+    user.gds_editor?
+  end
+end
```
Should this class have a spec? (I realise it's moved from somewhere else and didn't have coverage there either)

christophwong replied:
Thanks for pointing this out, I'm adding it in a newer commit.

Commit:
Unit tests for document policy

The design of this test is to focus on revealing what the 3 roles mean
in code, for example, a writer is a user without editor or gds_editor in
their permissions attribute. This is because the source code for policy
classes documented what permissions are given to which role, but it is
not obvious what is being checked when we say we are checking for a
departmental editor.

Commit:
Point all manual related controllers to use manual policy

Removes unnecessary policies for manual section and manual section
attachment controllers because they have the same rules as manual

Commit:
Use pundit to replace scoping code in Manual#index action

This commit does not contain test code but there is an existing feature
test

h-lame commented on spec/policies/document_policy_spec.rb:
```diff
@@ -0,0 +1,40 @@
+
+    it 'grants access to users from the organisation to which the document belongs' do
+      expect(described_class).to permit(cma_writer, CmaCase)
+    end
```
Without wishing to bikeshed could we make these more explicit by using spec-local `Document` subclasses that explicitly have different sets of organisations in them?  To understand this spec I have to go and look at the implementations of `AaibReport` and `CmaCase` to see that they have different organisations - if they were spec-local it would all be right here.

h-lame replied:
Have reworked the spec for this in a separate commit.

h-lame commented on app/policies/manual_policy.rb:
```diff
@@ -0,0 +1,28 @@
+  alias_method :update?, :index?
+  alias_method :show?, :index?
+
+  class Scope
```
I know that the commit message for adding this `Scope` says no tests were included because it's a refactoring already covered by a feature spec, and I mostly agree with the sentiment.  But... this scope is fairly important and to have the documentation for how it works hidden in a feature spec seems a little opaque so I'm somewhat conflicted.  Anyone else have thoughts?

h-lame replied:
Have added a spec for this in a separate commit.

h-lame committed:
Make document policy specs more explict

We use local document classes that explicitly state which orgs they allow
so that we don't need to look at other files to understand what is going
on.

h-lame committed:
Provide spec for ManualPolicy::Scope

This spec uses mocks which could be brittle, the alternative is stubbing
lots of api requests though which is no less brittle.
